test/tuple.jl

```diff
@@ -28,11 +28,8 @@ end
     @test convert(Tuple{Int, Int, Float64}, (1, 2, 3)) === (1, 2, 3.0)
 
     @test convert(Tuple{Float64, Int, UInt8}, (1.0, 2, 0x3)) === (1.0, 2, 0x3)
-    @test convert(NTuple, (1.0, 2, 0x3)) === (1.0, 2, 0x3)
     @test convert(Tuple{Vararg{Int}}, (1.0, 2, 0x3)) === (1, 2, 3)
     @test convert(Tuple{Int, Vararg{Int}}, (1.0, 2, 0x3)) === (1, 2, 3)
-    @test convert(Tuple{Vararg{T}} where T<:Integer, (1.0, 2, 0x3)) === (1, 2, 0x3)
```

Comment:
`Tuple{Vararg{Integer}}`?

Reply:
Good point. In fact we need a new method to handle that now.

Reply:
So, one downside to making the element type inside `Vararg` diagonal is that `Tuple{Vararg{Integer}}` doesn't match `Tuple{Vararg{T}} where T`. In fact I can't find a type that includes all such vararg types without also including `NTuple`, which you can't sensibly convert to. AFAICT, the best we can do is adding a type assertion to the conversion method to make sure the result is valid.

Reply:
Arguably, I'm not convinced the following normalization is correct:
```
julia> Tuple{Vararg{T where T<:S}} where S
Tuple{Vararg{S,N} where N} where S
```

Reply:
Fair point; I would put all the blame on the diagonal rule however --- it's too structure-based, so potentially all sorts of rearrangements get in its way.